### ISSUE TITLE
Improve Cluster join API. Join should be async

### ISSUE THREAD

**Issue by myroslavlisniak** [scalecube/scalecube#28]
Cluster join API should be async.
Instead of :
```java
ICluster join();
```
It should be something like this: 
```java
ListenableFuture<ICluster> join();
```
or if we don't need to return clutser it can be future of void
```java
ListenableFuture<Void> join();
```

**Comment by antonkharenko** [comment]
In what case returning ICluster in a future may be useful?

**Comment by myroslavlisniak** [comment]
I think returning ICluster is better, because we can write function (callbacks) without capturing of outer variables. Possible use case is send something to the cluster after joining or even start listening  for incoming messages.

**Comment by myroslavlisniak** [comment]
@antonkharenko Can you remind me why we need async join?

**Comment by antonkharenko** [comment]
@myroslavlisniak Actually the main reason is to be consistent with our principles and reactive approach which in a simplified form may be told as "Never block!". Also cluster join is long running operation and it involves several network messages to be sent and received. Also you can always make block on the future so the application calling the join may either block and wait join to be finished or register async callback, but currently there is no choice other then to be blocked on join.

**Comment by myroslavlisniak** [comment]
Merged #46
